src/modes/ModeEstimation.jl

```diff
@@ -0,0 +1,235 @@
+module ModeEstimation
+
+using ..Turing
+using ..Bijectors
+using LinearAlgebra
+
+import ..DynamicPPL
+import Optim
+import NamedArrays
+import ..ForwardDiff
+
+export MAP, MLE
+
+"""
+    ModeResult{
+        V<:NamedArrays.NamedArray, 
+        M<:NamedArrays.NamedArray, 
+        O<:Optim.MultivariateOptimizationResults, 
+        S<:NamedArrays.NamedArray
+    }
+
+A wrapper struct to store various results from a MAP or MLE estimation.
+
+Fields:
+
+- `values` is a vector with the resulting point estimates
+- `info_matrix` is the inverse Hessian
+- `optim_result` is the stored Optim.jl results
+- `summary_table` is a summary table with parameters, standard errors, and 
+  t-statistics computed from the information matrix.
+- `lp` is the final likelihood.
+"""
+struct ModeResult{
+    V<:NamedArrays.NamedArray, 
+    M<:Union{Missing, NamedArrays.NamedArray}, 
+    O<:Optim.MultivariateOptimizationResults, 
+    S<:NamedArrays.NamedArray
+}
+    values :: V
+    info_matrix :: M
+    optim_result :: O
+    summary_table :: S
+    lp :: Float64
+end
+
+function Base.show(io::IO, m::ModeResult)
+    show(io, m.summary_table)
+end
+
+"""
+    make_logjoint(model::DynamicPPL.Model, ctx::DynamicPPL.AbstractContext)
+
+Constructs a log density function that accepts a vector `z` and returns 
+a tuple (-likelihood, `varinfo`). The model is run using the provided 
+context `ctx`.
```

Comment:
The [first one-line sentence in each docstring should use the imperative form](https://docs.julialang.org/en/latest/manual/documentation/):
```suggestion
Construct a log density function.

The `model` is run using the provided context `ctx`.
```